docs/design-proposals/global-scheduler/GlobalSchedulerArchitectureDesignSpecificationFirstDraft.md

```diff
@@ -0,0 +1,416 @@
+    <td>
+      <li>Distribute data store for high availability</li>
+      <li>Support Data Persistency</li>
+      <li>Support both strong and eventual data consistency</li>
+      <li>Support multiple partitions for scalability</li>
+      <li>Notify object changes to API server which will notify the VM/Container Request Distributor component</li>
+    </td>
+  </tr>
+  
+  <tr>
+    <td>Dynamic Cluster Resource Info Cache</td>
+    <td>Distributed database cache to store the Cluster's available resource capacity. Since this information changes dynamically and will be collected periodically, there is no need to persist this information. Either ETCD Cache or Ignite satisfies our need for scalability, high availability, and cache performance<br><br>Each Global Scheduler partition's workload, geolocation tag, and profile tags are also saved in this DB cache</td>
+    <td>
+      <li>Distribute data store for scalability and high availability</li>
+      <li>Support cache data store for high performance</li>
+      <li>Support fast read operation</li>
+      <li>Support eventual data consistency</li>
+      <li>Notify object changes to API server which will notify the VM/Container Request Distributor</li>
+    </td>
+  </tr>
+  
+  <tr>
+    <td>VM/Container Request Distributor</td>
+    <td>The Distributor tags each incoming VM/Container request based on its specification of geolocation/region/AZ and resource profile. It then assigns a Home Global Scheduler to it based on tag matching (each Global Scheduler is also tagged based on its assigned geolocations/regions/AZs and resource profiles supported by those geolocation/regions/AZs). If there are two or more matching Global Schedulers, the Distributor assigns a Global Scheduler that has the lowest load to it.</td>
+    <td>
+      <li>Register with API server to get notification of new or changes of VM/Container object</li>
+      <li>Distribute the VM/Container scheduling requests to an appropriate Home Global Scheduler</li>
+      <li>Support multiple partitions for scalability. If multiple partitions are needed, incoming requests from the API servers will be load balanced to those Distributor partitions. That is, each Distributor will be connected to a subset of API Servers and will handle a subset of incoming requests</li>
+      <li>Each Distributor is connected to all Global Schedulers and will distribute an incoming VM/Container request to a Home Global Scheduler based on tag matching</li>
+    </td>
+  </tr>
+  
+  <tr>
+    <td>Dynamic Partition Mapper</td>
```

Comment:
1. Check whether each cluster is allocated to only one partition and scheduled by only one scheduler.
2. If the current partition scheduling mode is used, is batch scheduling not supported? For example, a resource request for batch scheduling may need to apply for resources in multiple partitions based on geographical distribution, which cannot be met by a single partition or a single scheduler.

Reply:
I have not got time to think a lot about batch scheduling. Originally we only consider support for replicaset scheduling. For batch scheduling, we need to first define if batch scheduling requires atomic operation or not. If the requests in the batch has different profiles, then they will be distributed to different schedulers (or schediuling partitions).  If the batch needs atomic operation, i.e. either all succeed or all fail, then the batch object status should not be updated until all requests associated with this batch is received. Some structure outside these schedulers must track the scheduling result of all the batch requests and do the update. In implementation, we can probably put this tracking functionality inside the distributor.

Reply:
Yes, I think our batch scenario, needs atomic operation.